src/Microsoft.OData.Client/BulkUpdateSaveResult.cs

```diff
@@ -202,19 +275,350 @@ protected override ODataRequestMessageWrapper CreateRequestMessage(string method
+                headers: null, 
+                statusCode: -1, 
+                this.operationResponses, 
+                batchResponse: false);
+            Descriptor descriptor = null;
+
+            try
+            {
+                if (this.adjacentDescriptors.Count == 0)
+                {
+                    descriptor = this.bulkUpdateGraph.TopLevelDescriptors[0];
+                    this.adjacentDescriptors.Push((0, this.bulkUpdateGraph.TopLevelDescriptors));
+                }
+                else
+                {
+                    Descriptor lastVisitedDescriptor = this.GetLastVisitedDescriptor();
+
+                    if (lastVisitedDescriptor != null)
+                    {
+                        IReadOnlyList<Descriptor> relatedDescriptors = this.bulkUpdateGraph.GetRelatedDescriptors(lastVisitedDescriptor);
+
+                        if (relatedDescriptors.Count == 0)
+                        {
+                            if (this.adjacentDescriptors.Peek().Item1 == this.adjacentDescriptors.Peek().Item2.Count - 1)
+                            {
+                                this.adjacentDescriptors.Pop();
+                            }
+
+                            int index = this.adjacentDescriptors.Peek().Item1;
+                            IReadOnlyList<Descriptor> descriptors = this.adjacentDescriptors.Peek().Item2;
+
+                            if (index == descriptors.Count - 1)
+                            {
+                                this.adjacentDescriptors.Pop();
+                            }
+                            else
+                            {
+                                descriptor = descriptors[index + 1];
+                                this.adjacentDescriptors.Pop();
+                                this.adjacentDescriptors.Push((index + 1, descriptors));
+                            }
+                        }
+                        else 
+                        {
+                            descriptor = relatedDescriptors[0];
+                            this.adjacentDescriptors.Push((0, relatedDescriptors));
+                        }
+                    }          
+                }
+                 
+                if (descriptor != null) 
+                {
+                    this.SaveResultProcessed(descriptor, HasFailedOperation(materializerState));
+
+                    operationResponse = new ChangeOperationResponse(this.cachedResponse.Headers, descriptor)
+                    {
+                        StatusCode = (int)this.cachedResponse.StatusCode
+                    };
+
+                    this.materializerStateForDescriptor.Add(descriptor, materializerState);
+
+                    if (this.cachedResponse.Exception != null)
+                    {
+                        operationResponse.Error = this.cachedResponse.Exception;
+
+                        if (ex == null)

```

Comment:
Given that this line is not in a loop, can `ex` have an other value apart from `null`